### PR TITLE
feat(chat actions): add reaction to message

### PR DESCRIPTION
**Pull request by mthinh**
# What ? 
- Implement add emoji reaction to message

### CODE REVIEW

File: lib/core/presentation/pages/chat/chat_message/chat_message_page.dart [hunk 1]
```diff
@@ -250,6 +250,28 @@ class ChatController extends State<ChatPageWithRoom> {
...
+          (event) => event.senderId == event.room.client.userID && event.type == 'm.reaction',
+        );
+    // prevent duplicated reactions
+    bool reacted = _allReactionEvents.any(
+      (e) => e.content.tryGetMap('m.relates_to')?['key'] == emoji,
+    );
+    if (reacted) {
+      return;
+    }
+    await  room.sendReaction(
```
Comment (sieuhuflit) [inline lib/core/presentation/pages/chat/chat_message/chat_message_page.dart:269]: code format quite weird

File: lib/theme/sizing.dart [hunk 2]
```diff
@@ -1,15 +1,22 @@
 class Sizing {
+  /// represent size of 8.0
```
Comment (sieuhuflit) [inline lib/theme/sizing.dart:2]: Why not Spacing and define this Sizing

**Review by sieuhuflit — APPROVED** [review]
Look good, leaving some comments, also better provide gif or image demo

File: lib/theme/sizing.dart (hunk 2 above)
Comment (RonTran636) [reply lib/theme/sizing.dart:2]: We don't need to define Sizing class just serves the same purpose as Spacing. Remove all the variables that we haven't used yet to clarify.

File: lib/core/presentation/pages/chat/chat_message/widgets/message_item/message_reaction_widget.dart [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+          horizontal: Spacing.superExtraSmall,
+          vertical: Spacing.superExtraSmall / 2,
+        ),
+        child: _buildContent(),
+      ),
+    );
+  }
+
+  Widget _buildContent() {
+    var renderKey = Characters(reactionKey!);
```
Comment (RonTran636) [inline lib/core/presentation/pages/chat/chat_message/widgets/message_item/message_reaction_widget.dart:161]: Boilerplate codes, put it into main `build` class. We don't need to separate it unless it very complicate that need to be separate

File: lib/core/presentation/pages/chat/chat_message/widgets/message_item/message_reaction_widget.dart [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+import 'package:app/injection/register_module.dart';
+import 'package:app/theme/color.dart';
+import 'package:app/theme/sizing.dart';
+import 'package:app/theme/spacing.dart';
+import 'package:flutter/material.dart';
+import 'package:matrix/matrix.dart';
+import 'package:collection/collection.dart' show IterableExtension;
+
+class MessageReactions extends StatelessWidget {
+  final Event event;
```
Comment (RonTran636) [inline lib/core/presentation/pages/chat/chat_message/widgets/message_item/message_reaction_widget.dart:14]: Add `widget` suffix

File: lib/core/presentation/pages/chat/chat_message/widgets/message_item/message_reaction_widget.dart [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+        reactionMap[key]!.reactors!.add(e.senderFromMemoryOrFallback);
+        reactionMap[key]!.reacted |= e.senderId == e.room.client.userID;
+      }
+    }
+    return reactionMap.values.toList();
+  }
+
+  _onTapReaction(_ReactionEntry r) async {
+    if (r.reacted) {
+      final evt = getAllReactionEvents().firstWhereOrNull(
```
Comment (RonTran636) [inline lib/core/presentation/pages/chat/chat_message/widgets/message_item/message_reaction_widget.dart:53]: Add Future\<void\>

**Review by RonTran636 — CHANGES_REQUESTED** [review]

**Comment by RonTran636** [comment]
@mthinh 
Avoid the use Helper Method when facing a complex UI widget. Create a new separate widget instead.
Here the referral you might need: https://www.youtube.com/watch?v=IOyq-eTRhvo

File: lib/theme/sizing.dart (hunk 2 above)
Comment (mthinh) [reply lib/theme/sizing.dart:2]: @RonTran636 @sieuhuflit Sizing definitely has another meaning with Spacing, even if some of their value might be the same with Spacing, but the meaning is different. FYI

File: lib/core/presentation/pages/chat/chat_message/widgets/message_item/message_reaction_widget.dart (hunk 4 above)
Comment (mthinh) [reply lib/core/presentation/pages/chat/chat_message/widgets/message_item/message_reaction_widget.dart:14]: @sieuhuflit may need your opinion on this ? 
We already had Bloc | Page | Utils as a suffix for bloc and page and other utils. Then the rest obviously widget, can consider if we still need Widget as a suffix ? 
Personally I prefer naming without Widget.

**Comment by mthinh** [comment]
> 

that sounds good, thanks for sharing

**Comment by mthinh** [comment]
> @mthinh Avoid the use Helper Method when facing a complex UI widget. Create a new separate widget instead. Here the referral you might need: https://www.youtube.com/watch?v=IOyq-eTRhvo

That sounds good, thanks for sharing

**Review by RonTran636 — APPROVED** [review]

**Pull request merged**